File: MyCsharpBot/UpdateSpPricesMarketBot.cs
// Bfexplorer cannot be held responsible for any losses or damages incurred during the use of this betfair bot.
// It is up to you to determine the level of risk you wish to trade under. 
// Do not gamble with money you cannot afford to lose.

using BeloSoft.Bfexplorer.Domain;
using BeloSoft.Bfexplorer.Service;
using BeloSoft.Bfexplorer.Trading;
using Microsoft.FSharp.Core;
using System;
using ApiModels = BeloSoft.Betfair.API.Models;

namespace MyCsharpBot
{
    /// <summary>
    /// BetfairSpPrices
    /// </summary>
    public class BetfairSpPrices
    {
        public double NearPrice { get; set; }
        public double FarPrice { get; set; }
        public double ActualSP { get; set; }
    }

    /// <summary>
    /// UpdateSpPricesBotParameters
    /// </summary>
    [Serializable]
    public sealed class UpdateSpPricesBotParameters : BotParameters
    {
        public UpdateSpPricesBotParameters()
        {
            UpdateInterval = TimeSpan.FromSeconds(5);
        }

        public TimeSpan UpdateInterval { get; set; }
    }

    /// <summary>
    /// SelectionUpdateFunc
    /// </summary>
    public abstract class SelectionUpdateFunc : FSharpFunc<Tuple<Selection, ApiModels.Runner>, Unit>
    {
    }

    /// <summary>
    /// UpdateSelectionSpPricesFunc
    /// </summary>
    public sealed class UpdateSelectionSpPricesFunc : SelectionUpdateFunc
    {
        public override Unit Invoke(Tuple<Selection, ApiModels.Runner> tupledArg)
        {
            Selection selection = tupledArg.Item1;
            ApiModels.Runner runner = tupledArg.Item2;

            BetfairSpPrices betfairSpPrices = selection.Data["betfairSpPrices"] as BetfairSpPrices;
            ApiModels.StartingPrices spPrices = runner.sp;

            betfairSpPrices.NearPrice = spPrices.nearPrice;
            betfairSpPrices.FarPrice = spPrices.farPrice;
            betfairSpPrices.ActualSP = spPrices.actualSP;

            return null;
        }
    }

    /// <summary>
    /// DataUpdateFunc
    /// </summary>
    public abstract class DataUpdateFunc : FSharpFunc<bool, Unit>
    {
    }

    /// <summary>
    /// UpdateMarketDataFunc
    /// </summary>
    public sealed class UpdateMarketDataFunc : DataUpdateFunc
    {
        public override Unit Invoke(bool result)
        {
            return null;
        }
    }

    /// <summary>
    /// UpdateSpPricesMarketBot
    /// </summary>
    public sealed class UpdateSpPricesMarketBot : MarketBot
    {
        // Const
        private static ApiModels.PriceProjection BetfairSpPriceProjection = ApiModels.PriceProjection.DefaultBetfairSpPrice();

        private static UpdateSelectionSpPricesFunc updateSelectionSpPricesFunc = new UpdateSelectionSpPricesFunc();
        private static UpdateMarketDataFunc updateMarketDataFunc = new UpdateMarketDataFunc();

        // Data
        private UpdateSpPricesBotParameters myParameters;
        private DateTime timeToUpdated = DateTime.MinValue;

        public UpdateSpPricesMarketBot(Market market, UpdateSpPricesBotParameters parameters, IBfexplorerService bfexplorerService)
            : base(market, parameters, bfexplorerService, FSharpOption<ICriteriaEvaluator>.None)
        {
            myParameters = parameters;

            foreach (var selection in market.Selections)
            {
                selection.Data["betfairSpPrices"] = new BetfairSpPrices();
            }
        }

        public override void Execute()
        {
            switch (this.Status)
            {
                case BotStatus.WaitingForEntryCriteria:
                    if (this.EvaluateCriteria())
                    {
                        this.Status = BotStatus.WaitingForOperation;
                    }
                    break;
                case BotStatus.WaitingForOperation:

                    if (DateTime.Now >= timeToUpdated)
                    {
                        this.UpdateMarketData(BetfairSpPriceProjection, updateSelectionSpPricesFunc, updateMarketDataFunc);
                        timeToUpdated = DateTime.Now.Add(myParameters.UpdateInterval);
                    }

                    if (this.RunningOnMarket.IsInPlay)
                    {
                        this.Status = BotStatus.ExecutionEnded;
                    }
                    break;
                default:
                    break;
            }
        }
    }
}

File: MyFsharpBot/UpdateSpPricesMarketBot.fs
// Bfexplorer cannot be held responsible for any losses or damages incurred during the use of this betfair bot.
// It is up to you to determine the level of risk you wish to trade under. 
// Do not gamble with money you cannot afford to lose.

namespace MyFsharpBot

open System

open BeloSoft.Betfair
open BeloSoft.Bfexplorer.Domain
open BeloSoft.Bfexplorer.Service
open BeloSoft.Bfexplorer.Trading

/// <summary>
/// BetfairSpPrices
/// </summary>
type BetfairSpPrices() =
    member val NearPrice : float = 0.0 with get, set
    member val FarPrice : float = 0.0 with get, set
    member val ActualSP : float = 0.0 with get, set

/// <summary>
/// UpdateSpPricesBotParameters
/// </summary>
[<Sealed>]
type UpdateSpPricesBotParameters() =
    inherit BotParameters()
    
    member val UpdateInterval : TimeSpan = TimeSpan.FromSeconds(5.0) with get, set

module MyBotData =

    let BetfairSpPriceProjection = API.Models.PriceProjection.DefaultBetfairSpPrice()

    let UpdateSelectionSpPrices(selection : Selection, runner : API.Models.Runner) =
        let betfairSpPrices = selection.Data.["bsp"] :?> BetfairSpPrices
        let spPrices = runner.sp

        betfairSpPrices.NearPrice <- spPrices.nearPrice
        betfairSpPrices.FarPrice <- spPrices.farPrice
        betfairSpPrices.ActualSP <- spPrices.actualSP

/// <summary>
/// UpdateSpPricesMarketBot
/// </summary>
[<Sealed>]
type UpdateSpPricesMarketBot(market : Market, parameters : UpdateSpPricesBotParameters, bfexplorerService : IBfexplorerService) as this =
    inherit MarketBot(market, parameters, bfexplorerService)

    let mutable timeToUpdated = DateTime.MinValue

    let isBetfairSpSettled() =
        let selection = market.Selections |> Seq.find isActiveSelection
        
        (selection.Data.["bsp"] :?> BetfairSpPrices).ActualSP <> 0.0

    do
        this.Status <- BotStatus.WaitingForEntryCriteria

        market.Selections
        |> Seq.iter (fun selection -> selection.Data.["bsp"] <- BetfairSpPrices())

    override this.Execute() =
        match this.Status with
        | BotStatus.WaitingForEntryCriteria -> 

            if this.EvaluateCriteria()
            then
                this.Status <- BotStatus.WaitingForOperation

        | BotStatus.WaitingForOperation ->

            if DateTime.Now >= timeToUpdated
            then
                this.UpdateMarketData(MyBotData.BetfairSpPriceProjection, MyBotData.UpdateSelectionSpPrices, fun _ -> ())
                timeToUpdated <- DateTime.Now.Add(parameters.UpdateInterval)
            
            if market.IsInPlay && isBetfairSpSettled()
            then                
                this.Status <- BotStatus.ExecutionEnded

        | _ -> ()